Add extent interaction

tbarsballe opened the pull request:
As described in https://github.com/openlayers/ol3/issues/5283

ahocevar commented on src/ol/interaction/extentinteraction.js:
```diff
@@ -0,0 +1,429 @@
+   * @this {ol.interaction.Extent}
+   * @private
+   */
+  this.handleEvent = function(mapBrowserEvent) {
```
To match the coding style we use for other components, can you please define all member methods outside the constructor, i.e. `ol.interaction.Extent.prototype.handleEvent = function(mapBrowserEvent) ...`.

ahocevar commented on src/ol/interaction/extentinteraction.js:
```diff
@@ -0,0 +1,429 @@
+  });
+
+  /* Other Callbacks */
+  var calculateExtent = function(point1, point2) {
```
You can use `ol.extent.boundingExtent([point1, point2])` instead.

ahocevar commented on src/ol/interaction/extentinteraction.js:
```diff
@@ -0,0 +1,429 @@
+ol.interaction.Extent.prototype.setMap = function(map) {
+  this.bboxOverlay_.setMap(map);
+  this.vertexOverlay_.setMap(map);
+  ol.interaction.Interaction.prototype.setMap.call(this, map);
```
Use `goog.base(this, 'setMap', map);` instead.

ahocevar commented on src/ol/interaction/extentinteraction.js:
```diff
@@ -0,0 +1,429 @@
+ * @param {ol.proj.Projection=} projection Projection to retun the extent in
+ * @return {ol.Extent} Drawn extent in the passed projection, or in the native projection if undefined.
+ */
+ol.interaction.Extent.prototype.getExtent = function(projection) {
```
Looks like all methods from here on could be `@private`.

tbarsballe replied:
setExtent is also intended to be public.

Sample use case is if you have an extent that you can set by text field and this interaction, you want each to be able to update the other.

ahocevar commented on src/ol/interaction/extentinteraction.js:
```diff
@@ -0,0 +1,429 @@
+      return ol.coordinate.squaredDistanceToSegment(pixelCoordinate, a) -
+          ol.coordinate.squaredDistanceToSegment(pixelCoordinate, b);
+    };
+    var getSegments = function(extent) {
```
You can use `ol.geom.Polygon.fromExtent` here.

tbarsballe replied:
No, I can't. This returns a list of line segments.

ahocevar commented on src/ol/interaction/extentinteraction.js:
```diff
@@ -0,0 +1,429 @@
+ * @extends {ol.events.Event}
+ */
+ol.interaction.ExtentEvent = function() {
+  goog.base(this, 'bbox');
```
For convenience, it would make sense to set the extent on the event object.

And for consistency with other events, it would make sense to define `'bbox'` as ExtentEventType.

tbarsballe replied:
> For convenience, it would make sense to set the extent on the event object.

Not sure what you mean by this? If you mean that the event should be responsible for calling setExtent, that makes having setExtent as a public mehtod problematic (see above)

ahocevar replied:
I meant you could give the event an `extent` property, so you could do something like
```js
extentInteraction.on('bbox', function(e) {
  console.log(e.extent.toString();
});
```
This also reminds me that `'bbox'` is probably not a good event name. Maybe `'modify'`?

tbarsballe replied:
Ah, that makes sense. Will do.

ahocevar commented on externs/olx.js:
```diff
@@ -2734,6 +2734,17 @@ olx.interaction.DrawOptions.prototype.wrapX;
 
 /**
+ * @typedef {{extent: (ol.Extent|undefined),
+ *     extentProjection: (ol.proj.ProjectionLike|undefined),
```
Name could just be `projection`.

ahocevar commented on externs/olx.js:
```diff
@@ -2734,6 +2734,17 @@ olx.interaction.DrawOptions.prototype.wrapX;
 /**
+ * @typedef {{extent: (ol.Extent|undefined),
+ *     extentProjection: (ol.proj.ProjectionLike|undefined),
+ *     bboxStyle: (ol.style.Style|Array.<ol.style.Style>|ol.style.StyleFunction|undefined),
```
Name could be `boxStyle` or just `style`.

ahocevar commented on externs/olx.js:
```diff
@@ -2734,6 +2734,17 @@ olx.interaction.DrawOptions.prototype.wrapX;
+ *     wrapX: (boolean|undefined)}}
+ * @api
+ */
+olx.interaction.ExtentOptions;
```
In addition to this typedef, you'll need to define each option as `@type` below (see other options in `externs/olx.js`).

ahocevar commented on src/ol/interaction/extentinteraction.js:
```diff
@@ -0,0 +1,429 @@
+ *
+ * @param {function (?): ?} cb Event callback
+ */
+ol.interaction.Extent.prototype.addExtentListener = function(cb) {
```
Probably not needed - users can use `ol.interaction.Extent#on()`.

ahocevar commented on src/ol/interaction/extentinteraction.js:
```diff
@@ -0,0 +1,429 @@
+ *
+ * @param {function (?): ?} cb Event callback
+ */
+ol.interaction.Extent.prototype.removeExtentListener = function(cb) {
```
Probably not needed - users can use `ol.interaction.Extent#un()`.

ahocevar commented on src/ol/interaction/extentinteraction.js:
```diff
@@ -0,0 +1,429 @@
+ * @param {ol.Extent} extent Extent
+ * @param {ol.proj.Projection|string=} projection Projection to return the extent in
+ */
+ol.interaction.Extent.prototype.setExtent = function(extent, projection) {
```
If you decide to make this `@api`, I'd go without the `projection` argument here and the `projection` constructor option, and use the view projection. Otherwise it will be confusing for users, because we do not do such thing elsewhere.

tbarsballe replied:
Fair enough. The projection stuff is all carry-over from the origional use case, which had the view in EPSG:3857 but displayed the extent values in lat/lon. I will remove the projection stuff from this.

ahocevar commented:
Thanks @tbarsballe for this pull request. I added a few remarks. Let me know if you have questions.

Would you be able to add a few unit tests? As you can see, our automated tests currently fail when the test coverage drops by more than 0.5%.

marcjansen commented:
Very nice work! And exceptional nice review, @ahocevar 

👌👏👏

tbarsballe commented:
I made the changes you suggested @ahocevar (except where otherwise noted in the discussion above) and added test cases.

ahocevar commented on src/ol/interaction/extentinteraction.js:
```diff
@@ -0,0 +1,460 @@
+  var bboxFeature = this.bboxFeature_;
+
+  if (!extent) {
+    extent = [0,0,0,0];
```
You should be able to use `ol.extent.createEmpty();` here.

ahocevar commented on src/ol/interaction/extentinteraction.js:
```diff
@@ -0,0 +1,460 @@
+ * @private
+ */
+ol.interaction.Extent.prototype.snapToVertex = function(pixel, map) {
+  var pixelCoordinate = map.getCoordinateFromPixel(pixel);
```
I'm pretty sure you could simplify this method to
```js
return this.extent_ ?
    this.bboxFeature_.getGeometry().closestPointXY(map.getCoordinateFromPixel(pixel)) :
    null;`
```

ahocevar replied:
Oh, I see that you only snap within a threshold. But still, you can combine the above with `ol.extent.closestSquaredDistanceXY` to determine whether you're within the threshold.

tbarsballe replied:
This one I adapted straight from modifyinteraction.js. I only snap within a threshold. Additionally, if I am close to a corner, I snap to the vertex rather than the edge.

ahocevar commented on src/ol/interaction/extentinteraction.js:
```diff
@@ -0,0 +1,460 @@
+    extent = [0,0,0,0];
+  }
+  if (!bboxFeature) {
+    bboxFeature = new ol.Feature(new ol.geom.Polygon([[
```
What was the reason for not using `ol.geom.Polygon.fromExtent(extent)` here?

tbarsballe replied:
I was unaware of that function. I will update.

ahocevar commented on src/ol/interaction/extentinteraction.js:
```diff
@@ -0,0 +1,460 @@
+    this.bboxOverlay_.getSource().addFeature(bboxFeature);
+  } else {
+    var geometry = (bboxFeature.getGeometry());
+    geometry.setCoordinates([[
```
You may want to use the opposite winding order here if you make the other changes I suggested.

ahocevar commented on src/ol/interaction/extentinteraction.js:
```diff
@@ -0,0 +1,460 @@
+/**
+ * @param {ol.MapBrowserEvent} mapBrowserEvent Event.
+ * @return {boolean} Propagate event?
+ * @this {ol.interaction.Extent}
```
No `@this` annotations should be necessary when you define methods like this.

ahocevar replied:
Ah, scratch that! These handlers are not supposed to be members, but static functions. So `ol.interaction.Extent.handleEvent` instead of `ol.interaction.Extent.prototype.handleEvent` below. Same for the other handlers you pass to the superclass constructor.

ahocevar commented on src/ol/interaction/extentinteraction.js:
```diff
@@ -0,0 +1,460 @@
+ * @this {ol.interaction.Extent}
+ * @private
+ */
+ol.interaction.Extent.prototype.handleDownEvent = function(mapBrowserEvent) {
```
For `@private` functions, we append `_` to the function name elsewhere.

ahocevar commented on src/ol/interaction/extentinteraction.js:
```diff
@@ -0,0 +1,460 @@
+ * @extends {ol.events.Event}
+ */
+ol.interaction.ExtentEvent = function(extent) {
+  goog.base(this, ol.interaction.ExtentEventType.EXTENTCHANGED);
```
You'll need to set the extent on the class, e.g.
```js
  /**
   * The current extent.
   * @type {ol.Extent}
   * @api
   */
  this.extent = extent;
```

tbarsballe replied:
That makes more sense. I thought I was missing something

ahocevar commented on src/ol/interaction/extentinteraction.js:
```diff
@@ -0,0 +1,460 @@
+  this.snappedToVertex_ = null;
+  this.bboxFeature_ = null;
+  this.vertexFeature_ = null;
+  this.bboxOverlay_ = null;
```
No need to initialize anything here that is initialized elsewhere in the constructor.

tbarsballe replied:
None of these are inititalized anywhere else in the constructor - I was getting linter warnings for all of them. These only get set in reaction to events, while the extent is active.

ahocevar replied:
`bboxOverlay_` and `vertexOverlay_` are defined a bit further down in the constructor. Note that all should get a `@type` annotation.

tbarsballe replied:
Ah, yes, you are right. I guess I was a bit overzealous with those two.

ahocevar commented on src/ol/interaction/extentinteraction.js:
```diff
@@ -0,0 +1,460 @@
+  this.pointerHandler = null;
+
+  this.pixelTolerance_ = 10;
+  this.lastPixel_ = null;
```
You should add `@type` (and `@private`, if appropriate) annotations for any member properties.

ahocevar commented on src/ol/interaction/extentinteraction.js:
```diff
@@ -0,0 +1,460 @@
+  }
+};
+
+/* Public Methods */
```
We do not add comments like this elsewhere.

ahocevar commented:
Thanks for your continued effort on this @tbarsballe. This is in good shape already. I added a few more comments, most about coding style deviations from what we use elsewhere, some suggestions for simplifications, and one about fixing the event being emitted.

probins commented:
this will have to be added to #5194

tbarsballe commented:
Next batch of fixes is added.
Changed event handlers to static
Added more docstrings to initializations
Added some function calls where applicable
Some code style fixup

probins commented on src/ol/interaction/extentinteraction.js:
```diff
@@ -0,0 +1,478 @@
+   */
+  EXTENTCHANGED: 'extentchanged'
+};
+/**
```
for consistency, there should be blank lines here

ahocevar commented on src/ol/interaction/extentinteraction.js:
```diff
@@ -0,0 +1,478 @@
+  return this.extent;
+};
+/**
+ * Manually sets the drawn extent, in the requested projection. Uses the native projection if undefined.
```
Docs need to be updated now that there is no `projection` argument any more.

ahocevar commented on src/ol/interaction/extentinteraction.js:
```diff
@@ -0,0 +1,478 @@
+  var extentFeature = this.extentFeature_;
+
+  if (!extent) {
+    extent = [0,0,0,0];
```
I'm not sure about this. What about creating a feature without geometry if `extent` is `null`?

tbarsballe replied:
That seemds like a better approach. Will update.

tbarsballe commented:
Next batch of fixes in.
* Added proper newlines
* Updated docs to remove mention of projection
* Use a feature without geometry if extent is null

Also, sorry for the extended delay getting these changes in, I suddenly ran out of time to work on this mid-May.

tbarsballe commented:
Rebased against master and updated to conform to changes introduced by #5194.

probins commented:
pls replace `goog.inherits` with `ol.inherits` as well.

tbarsballe commented:
Fixed

bartvde commented:
Having another look at this one now.

bartvde commented:
@tbarsballe any chance you can add an example to show off this new interaction? TIA

bartvde commented on src/ol/interaction/extent.js:
```diff
@@ -0,0 +1,506 @@
+
+/**
+ * @classdesc
+ * Events emitted by {@link ol.interaction.ExtentInteraction} instances are instances of
```
Should be ```{@link ol.interaction.Extent}``` now

bartvde commented on src/ol/interaction/extent.js:
```diff
@@ -0,0 +1,506 @@
+/**
+ * @classdesc
+ * Allows the user to draw a vector box by clicking and dragging on the map.
+ * Once drawn, the vector box can be modified by draging its vertices or edges.
```
dragging

bartvde commented on src/ol/interaction/extent.js:
```diff
@@ -0,0 +1,506 @@
+ *
+ * @constructor
+ * @extends {ol.interaction.Pointer}
+ * @fires ol.ExtentEvent
```
```ol.interaction.ExtentEvent```

bartvde commented on src/ol/interaction/extent.js:
```diff
@@ -0,0 +1,506 @@
+ * @param {olx.interaction.ExtentOptions} opts Options.
+ * @api
+ */
+ol.interaction.Extent = function(opts) {
```
very minor but we use options everywhere else

bartvde commented on src/ol/interaction/extent.js:
```diff
@@ -0,0 +1,506 @@
+   * @type {function (ol.Coordinate): ol.Extent|null}
+   * @private
+   */
+  this.pointerHandler = null;
```
private should be with _ at the end

bartvde commented on src/ol/interaction/extent.js:
```diff
@@ -0,0 +1,506 @@
+   * @type {ol.Extent}
+   * @private
+   */
+  this.extent = null;
```
private should be with _ at the end

bartvde commented on src/ol/interaction/extent.js:
```diff
@@ -0,0 +1,506 @@
+ol.interaction.Extent = function(opts) {
+
+  /**
+   * extent Extent
```
I'd take out the first line since it doesn't add much here, or change it to read one times extent only

bartvde commented on src/ol/interaction/extent.js:
```diff
@@ -0,0 +1,506 @@
+  this.extent = null;
+
+  /**
+   * pointerHandler_ Handler for mointer move events
```
just Handler for mointer move events, no need to repeat the name of the variable at the front

bartvde commented on src/ol/interaction/extent.js:
```diff
@@ -0,0 +1,506 @@
+  this.pointerHandler = null;
+
+  /**
+   * pixelTolerance_ pixel threshold to snap to extent
```
No need to have the variable name at the front

bartvde commented on src/ol/interaction/extent.js:
```diff
@@ -0,0 +1,506 @@
+  this.pixelTolerance_ = 10;
+
+  /**
+   * lastPixel_ Last known pixel coordinate of the pointer
```
No need to have the variable name at the front

bartvde commented on src/ol/interaction/extent.js:
```diff
@@ -0,0 +1,506 @@
+  this.lastPixel_ = null;
+
+  /**
+   * snappedToVertex_ Is the pointer snapped to an extent vertex
```
No need to have the variable name at the front

bartvde commented on src/ol/interaction/extent.js:
```diff
@@ -0,0 +1,506 @@
+  this.snappedToVertex_ = false;
+
+  /**
+   * extentFeature_ Feature for displaying the visible extent
```
No need to have the variable name at the front

bartvde commented on src/ol/interaction/extent.js:
```diff
@@ -0,0 +1,506 @@
+  this.extentFeature_ = null;
+
+  /**
+   * vertexFeature_ Feature for displaying the visible pointer
```
No need to have the variable name at the front

bartvde commented on src/ol/interaction/extent.js:
```diff
@@ -0,0 +1,506 @@
+   */
+  this.vertexFeature_ = null;
+
+  if (!opts) {
```
since it's optional, it needs to be renamed with the opt_ prefix

bartvde commented on src/ol/interaction/extent.js:
```diff
@@ -0,0 +1,506 @@
+  });
+
+  /**
+   * extentOverlay_ Layer for the extentFeature
```
No need to have the variable name at the front

bartvde commented on src/ol/interaction/extent.js:
```diff
@@ -0,0 +1,506 @@
+   * @type {ol.layer.Vector}
+   * @private
+   */
+  this.extentOverlay = new ol.layer.Vector({
```
missing _

bartvde commented on src/ol/interaction/extent.js:
```diff
@@ -0,0 +1,506 @@
+  });
+
+  /**
+   * vertexOverlay_ Layer for the vertexFeature
```
No need to have the variable name at the front

bartvde commented on src/ol/interaction/extent.js:
```diff
@@ -0,0 +1,506 @@
+ol.interaction.Extent.handleDragEvent_ = function(mapBrowserEvent) {
+  this.lastPixel_ = mapBrowserEvent.pixel;
+  if (this.pointerHandler) {
+    var pixelCoordinate = mapBrowserEvent.map.getCoordinateFromPixel(mapBrowserEvent.pixel);
```
why not use ```mapBrowserEvent.coordinate``` here?

tbarsballe replied:
Didn't read the doc / didn't know it was there. Updating.

bartvde commented on src/ol/interaction/extent.js:
```diff
@@ -0,0 +1,506 @@
+  this.pointerHandler = null;
+  //If bbox is zero area, set to null;
+  var extent = this.getExtent();
+  if (!extent || extent[0] == extent[2] || extent[1] == extent[3]) {
```
```!extent && ol.extent.getArea(extent) > 0``` ?

tbarsballe replied:
That is clearer, although the current one is computationally more efficient (though minor enough that it doesn't really matter).

bartvde commented on src/ol/interaction/extent.js:
```diff
@@ -0,0 +1,506 @@
+  var style = ol.style.Style.createDefaultEditing();
+  return function(feature, resolution) {
+    return style[ol.geom.GeometryType.POLYGON];
+  };
```
why do we need to create a style function here? Can't we just use the style object directly?

tbarsballe replied:
If something changes the default style mid application, I would assume we want everything to update. (Is changing the default style function something that is plausible to occur?)

bartvde replied:
No that is not possible since it's not in the API

bartvde commented on src/ol/interaction/extent.js:
```diff
@@ -0,0 +1,506 @@
+ */
+ol.interaction.Extent.getDefaultPointerStyleFunction_ = function() {
+  var style = ol.style.Style.createDefaultEditing();
+  return function(feature, resolution) {
```
why do we need to create a style function here? Can't we just use the style object directly?

bartvde commented on src/ol/interaction/extent.js:
```diff
@@ -0,0 +1,506 @@
+ * @returns {function (ol.Coordinate): ol.Extent} event handler
+ * @private
+ */
+ol.interaction.Extent.prototype.getPointHandler_ = function(fixedPoint) {
```
can this be static instead?

bartvde commented on src/ol/interaction/extent.js:
```diff
@@ -0,0 +1,506 @@
+ * @returns {function (ol.Coordinate): ol.Extent|null} event handler
+ * @private
+ */
+ol.interaction.Extent.prototype.getEdgeHandler_ = function(fixedP1, fixedP2) {
```
can this be static instead?

tbarsballe replied:
No issue with making this static. Given that it is a private method, shouldn't really matter either way. I can update (I think - need to double chech how to do static JS/OL3 methods)

bartvde commented on src/ol/interaction/extent.js:
```diff
@@ -0,0 +1,506 @@
+ * @private
+ */
+ol.interaction.Extent.getSegments_ = function(extent) {
+  return [
```
could we use ```ol.geom.Polygon.fromExtent(extent).getCoordinates()``` instead?

tbarsballe replied:
Note that this is three levels of nested arrays (line secments rather than coordinates)

bartvde commented on src/ol/interaction/extent.js:
```diff
@@ -0,0 +1,506 @@
+    this.extentOverlay.getSource().addFeature(extentFeature);
+  } else {
+    if (!extent) {
+      extentFeature.setGeometry(null);
```
I wonder if this should this be undefined instead?

tbarsballe replied:
Undefined may be marginally more correct. Functionally, it will not make a difference.

bartvde commented on src/ol/interaction/extent.js:
```diff
@@ -0,0 +1,506 @@
+};
+
+/**
+ * Returns the current drawn extent in the native projection
```
with native do you mean the view projection?

tbarsballe replied:
Original implementation allowed for providing extents in arbitrary projections, so this is a bit of a holdover from that. But yes, native means the view projection.

bartvde commented on src/ol/interaction/extent.js:
```diff
@@ -0,0 +1,506 @@
+};
+
+/**
+ * Manually sets the drawn extent, useing the native projection.
```
using

bartvde commented on test/spec/ol/interaction/extent.test.js:
```diff
@@ -0,0 +1,157 @@
+  describe('draw extent', function() {
+
+    it('drawing extent works', function() {
+      var interaction = new ol.interaction.Extent();
```
these two lines could be in beforeEach?

bartvde commented:
thanks @tbarsballe I'm done with my review. If you have some time to adapt the PR we can get this in quickly. TIA.

tbarsballe commented:
Updated based on feedback. Example coming shortly.

tbarsballe commented:
Added extent interaction example.

bartvde commented:
Thanks @tbarsballe LGTM
